master/pkg/checkpoints/s3/s3checkpoints.go

```diff
@@ -105,13 +111,35 @@ func NewS3Downloader(
 
 	// We do not pass in credentials explicitly. Instead, we reply on
 	// the existing AWS credentials.
-	region, err := GetS3BucketRegion(ctx, bucket)
+	var region string
+	var err error
+	if endpointURL != nil {
+		endpointFormat := fmt.Sprint(*endpointURL, "/%s")
+		region, err = GetS3BucketRegion(ctx, bucket, &endpointFormat)
+	} else {
+		region, err = GetS3BucketRegion(ctx, bucket, nil)
+	}
+
 	if err != nil {
 		return nil, err
 	}
-	sess, err := session.NewSession(&aws.Config{
-		Region: &region,
-	})
+
+	var sess *session.Session
+	if endpointURL != nil {
+		sess, err = session.NewSession(&aws.Config{
+			Region:           aws.String(region),
+			Endpoint:         endpointURL,
+			DisableSSL:       aws.Bool(false),
+			S3ForcePathStyle: aws.Bool(true),
+		})
+		if err != nil {
+			return nil, err
```

Comment:
nit: we can probably drop this check in the if statement since we check again immediately after